app/models/pieces/piece.rb

```diff
@@ -59,16 +60,39 @@ def moving_diagonally?(x_destination, y_destination)
 
   # return false if attempting to move to current square
   # then determines if the destination square is within the board boundaries
+  # returns false if the destination is occupied and the moving piece and destination piece player_id's are the same
   # then (unless piece is a knight) determines if the piece is obstructed
   def valid_move?(x_destination, y_destination, knight = false)
     return false if x_destination == x_position && y_destination == y_position
     return false if !(0..7).cover?(x_destination) || !(0..7).cover?(y_destination)
+    return false if game.occupied?(x_destination, y_destination) && game.find_piece(x_destination, y_destination).player_id == player_id
```

Comment:
This performs the piece query twice by calling the two methods. If the player_id of the found piece is equal to player_id, wouldn't the square implicitly be occupied?